docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,895 @@
+field that holds configuration for a given resource identified by `UniqueResourceKey`. So when user creates policy that has 
+real resource in `spec.to[].targetRef` we will create confs by resources: (this is configuration that will be applied to selected dataplane)
+
+```yaml
+resourceRules:
+  backend.backend-ns:
+    conf: $conf1
+  backend-route.backend-ns:
+    conf: merge($conf1, $conf2)
+```
+
+Route configuration will be applied when route was created for a given DPP, if not MeshService config will be applied. 
+Route may not be present if `Mesh*Route` was applied only for subset of proxies, for example:
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: route-to-backend
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  targetRef:
+    kind: MeshSubset
+    tags:
+      k8s.kuma.io/namespace: frontend-ns
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: timout-for-backend
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  targetRef:
+    kind: MeshSubset
+    tags:
+      k8s.kuma.io/namespace: some-other-ns
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+```
+
+After applying this two policies timeout won't be applied to any proxy, since `Mesh*Route` and MeshTimeout is applied
+on different subset of proxies. We should ignore this situation as this will just have no effect. Also we don't have possibility
+to show warnings with this information to users. 
+
+#### Mesh*Route with multiple spec.to[].targetRef
+
+At the moment users can create `Mesh*Route` that references multiple MeshServices, like in example:
+
+```yaml
+kind: MeshHTTPRoute
+metadata:
+  name: route-1
+spec:
+  to:
+    - targetRef:
+        kind: MeshService
+        name: backend
+      rules: [...]
+    - targetRef:
+        kind: MeshService
+        name: frontend
+      rules: [...]
+```
+
+When we create MeshTimeout policy that configures `Mesh*Route` and each of `MeshService` when merging configuration we will
+have problem which conf should we pick for merging. With example timeout policy:
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTimeout
+metadata:
+  name: timeout-on-backend
+  namespace: backend-ns
+  labels:
+    kuma.io/mesh: default
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+    default: conf1
+  - targetRef:
+      kind: MeshService
+      name: frontend
+    default: conf2
+  - targetRef:
+      kind: MeshHTTPRoute
+      name: slow-backend-route
+      namespace: backend-ns
+    default: conf3
+```
+
+When merging we will merge `$conf1` and `$conf2` for MeshService with `$conf3` for `Mesh*Route`. 
+
+```yaml
+resourceRules:
+  backend.backend-ns: 
+    conf: $conf1
+  frontend.backend-ns:
+    conf: $conf2
+  backend-route.backend-ns:
+    conf: merge($conf1, $conf2, $conf3)
```

Comment:
Isn't there an index thing for this @lobkovilya ?

Reply:
You mean the [ruleIndex](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/060-policy-matching-with-real-resources.md?plain=1#L77) in origin?